File: routes/token.go
package routes

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/tiborv/prxy/db"
)

const tokenPath = "/api/token"

func init() {
	mux.Handle(tokenPath+"/list", RequireUser(listToken))
	mux.Handle(tokenPath+"/update", RequireUser(updateToken))
	mux.Handle(tokenPath+"/create", RequireUser(createToken))
	mux.Handle(tokenPath+"/delete", RequireUser(deleteToken))

}

func listToken(w http.ResponseWriter, r *http.Request) {
	tokens, _ := db.FindAllTokens()
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(tokens)
}

func createToken(w http.ResponseWriter, r *http.Request) {
	token, jsonErr := db.TokenJson(r.Body)
	if jsonErr != nil {
		fmt.Println("Token create jsonErr:", jsonErr)
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, "Token not created")
		return
	}
	savedToken, saveErr := token.Init().Save()
	if saveErr != nil {
		fmt.Println("Token create err:", saveErr)
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, "Token not created")
		return
	}
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(savedToken)
}

func updateToken(w http.ResponseWriter, r *http.Request) {
	token, jsonErr := db.TokenJson(r.Body)
	if jsonErr != nil {
		fmt.Println("Token update jsonErr:", jsonErr)
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, "Token not updated")
		return
	}
	savedToken, saveErr := token.Save()
	if saveErr != nil {
		fmt.Println("Token update err:", saveErr)
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, "Token not updated")
		return
	}
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(savedToken)
}

func deleteToken(w http.ResponseWriter, r *http.Request) {
	token, jsonErr := db.TokenJson(r.Body)
	if jsonErr != nil {
		fmt.Println("Token delete jsonErr:", jsonErr)
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprint(w, "Token not deleted")
		return
	}
	deleted := token.Delete()
	if deleted {
		w.WriteHeader(http.StatusOK)
		fmt.Fprint(w, "Token deleted")
		return
	}
	w.WriteHeader(http.StatusBadRequest)
	fmt.Fprint(w, "Token not deleted")

}

File: client/actions/token.js
import * as api from '../utils/token'

export const LIST_SUCESS = 'TOKEN.LIST_SUCESS'
export const list = () => (dispatch) => {
  api
    .list()
    .then(tokens => dispatch({
      type: LIST_SUCESS,
      tokens,
    }))
}

export const DELETE_SUCESS = 'TOKEN.DELETE_SUCESS'
export const del = (id) => (dispatch) => {
  api
    .del(id)
    .then(() => dispatch({
      type: DELETE_SUCESS,
      id
    }))
}

export const UPDATE_SUCESS = 'TOKEN.UPDATE_SUCESS'
export const update = (token) => (dispatch) => {
  api
    .update(token)
    .then(() => dispatch({
      type: UPDATE_SUCESS,
      token
    }))
}

export const CREATE_SUCESS = 'TOKEN.CREATE_SUCESS'
export const create = (token) => (dispatch) => {
  api
    .create(token)
    .then(() => dispatch({
      type: CREATE_SUCESS,
      token
    }))
}
